src/main/java/dev/sigstore/KeylessVerificationResult.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2022 The Sigstore Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package dev.sigstore;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import dev.sigstore.json.GsonSupplier;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class KeylessVerificationResult {
+
+  enum Check {
+    NONE("none"),
+    SIGNATURE("Signature is valid for artifact and public key"),
+    FULCIO_CERT("Fulcio certificate chains up to fulcio root"),
+    FULCIO_CERT_SCT("Fulcio certificate sct is signed by the ct log"),
+    REKOR_ENTRY_EXISTS("Rekor entry was found on the log"),
+    REKOR_ENTRY_SIGNED("Rekor entry is signed by the rekor public key"),
+    REKOR_ENTRY_CONTENTS("Rekor entry body content matches request"),
+    REKOR_ENTRY_INCLUSION("Rekor entry inclusion proof verified"),
+    SIGNATURE_TIME("Signing occurred while certificate was valid");
+
+    private final String description;
+
+    Check(String description) {
+      this.description = description;
+    }
+
+    public String getDescription() {
+      return description;
+    }
+  }
+
+  private final List<Check> passingChecks;
+  private final Check failedCheck;
+  private final String failedReason;
+
+  public KeylessVerificationResult(
+      ImmutableList<Check> passingChecks, Check failedCheck, String failedReason) {
+    this.passingChecks = passingChecks;
+    this.failedCheck = failedCheck;
+    this.failedReason = failedReason;
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static class Builder {
+    private final List<Check> passingChecks = new ArrayList<>();
+
+    private Builder() {}
+
+    public Builder checkPassed(Check check) {
+      if (passingChecks.contains(check)) {
+        throw new IllegalStateException(
```

Comment:
If I'm understanding the API correctly, I think this is IllegalArgumentException, because the caller should not place the check into the same list twice.  If we used a Set here it would be a no-op.

Reply:
I don't want it to be a no-op? Each check should only happen once. If there are multiple parts, it should be two checks.